Partner Experience: revamp horizon monitoring docs (#198)
stellar/stellar-docs: docs/run-platform-server/monitoring.mdx

## Metrics

> Metrics are emitted from Horizon over HTTP in [the de facto text based exposition format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format). The Metrics are published on the _private_ `/metrics` path of Horizon Admin port, which is an optional service to be started and will be bound by the horizon process onto the host machine loopback network(localhost or 127.0.0.1). To enable the Admin port, add environment configuration parameter `ADMIN_PORT=XXXXX`, the metrics endpoint will be reachable on the host machine as `localhost:<ADMIN_PORT>\metrics`. You can verify this by pointing any browser that can reach this address, it will print out all metrics keys.

Reviewer: ```suggestion
Metrics are emitted from Horizon over HTTP in [the de facto text based exposition format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format). The Metrics are published on the _private_ `/metrics` path of Horizon Admin port, which is an optional service to be started and will be bound by the horizon process onto the host machine loopback network(localhost or 127.0.0.1). To enable the Admin port, add environment configuration parameter `ADMIN_PORT=XXXXX`, the metrics endpoint will be reachable on the host machine as `localhost:<ADMIN_PORT>/metrics`. You can verify this by pointing any browser that can reach this address, it will print out all metrics keys.
```

### Exporting

Once the admin port is enabled, the Horizon metrics endpoint can be 'scraped' by external monitoring infrastructure. Since the metrics output is encoded to [standard text based format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format) it will be compatibile for usage with many types of monitoring infrastructure that interoperate with same standard format.

In the case of Horizon, metrics are published on the Admin http port which is bound to the host machine's loop back network interface(127.0.0.1), so external monitoring systems or services cannot reach the port directly. To expose the metrics securely, we recommend following the exporter pattern, which is common metrics scraping strategy.

> One real world example for exporting from a bare metal horizon installation(horizon has been installed directly onto an operating system), use[FluentBit and the Prometheus Exporter](https://docs.fluentbit.io/manual/pipeline/outputs/prometheus-exporter) on same host machine as Horizon is running. FluentBit will perform a simple port forwarding pipeline on host machine. Configure the input to be Horizon's `localhost:<ADMIN_PORT>\metrics` and the output to be the `host` and `port` representing the target network interface and port on the host machine, and then configure your monitoring infrastructure to scrape that host address.

Reviewer: ```suggestion
One real world example for exporting from a bare metal horizon installation (horizon has been installed directly onto an operating system), use [FluentBit and the Prometheus Exporter](https://docs.fluentbit.io/manual/pipeline/outputs/prometheus-exporter) on the same host machine as Horizon is running. FluentBit will perform a simple port forwarding pipeline on the host machine. Configure the input to be Horizon's `localhost:<ADMIN_PORT>/metrics` and the output to be the `host` and `port` representing the target network interface and port on the host machine, and then configure your monitoring infrastructure to scrape that host address.
```

> In container orchestrated environents such as kubernetes, you will use the same exporter strategy, and have already setup metrics infrastructure deployment on cluster such as Prometheus and Grafana via the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator), and will just need to configure that infrastructure to scrape the horizon pod based on the ADMIN_PORT.

Reviewer: ```suggestion
In container orchestrated environents such as kubernetes, you can use the same exporter strategy. We assume you already have a metrics infrastructure deployment like Prometheus and Grafana setup on the cluster via the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator), and will just need to configure that infrastructure to scrape the horizon pod based on the ADMIN_PORT.
```

### Data model

> There are numerous application metrics keys emitted by horizon at runtime, encoded in all four types of exposition formats: `counter`, `gauge`, `historgram`, `summary`. Each key is further qualified with labels for more granularity. To summarize, we can highlight the groupings of metrics keys by common function denoted in the prefix of their name:

Reviewer: ```suggestion
There are numerous application metrics keys emitted by horizon at runtime, encoded in four types of exposition formats: `counter`, `gauge`, `histogram`, `summary`. Each key is further qualified with labels for more granularity. To summarize, we can highlight the groupings of metrics keys by common function denoted in the prefix of their name:
```

> `go_` - golang specfici runtime performance `horizon_txsub_` - attributes of horizon transaction submission sub system if enabled. `horizon_stellar_core_` - runtime attributes of stellar network reported by the captive core or a remote stellar core which horizon is configured. `horizon_order_book_` - runtime attributes of the in memory order book maintained by horizon of the current stellar network `horizon_log_` - counters of how many log messages printed at each severity level `horizon_ingest_` - performance measurements and stateful aspects of horizon's internal ingestion sub system `horizon_http_` - statistics and measurements of horizon's HTTP API service, all aspects of request/response load and timings. `horizon_history_` - statistics on horizon ingested historical ledgers `horizon_db_` `process_` - generic host machine compute measurements

Reviewer: ```suggestion
- `go_`: golang specfici runtime performance 
- `horizon_txsub_`: attributes of horizon transaction submission sub system if enabled. 
- `horizon_stellar_core_`: runtime attributes of stellar network reported by the captive core or a remote stellar core which horizon is configured. 
- `horizon_order_book_`: runtime attributes of the in memory order book maintained by horizon of the current stellar network 
- `horizon_log_`: counters of how many log messages printed at each severity level 
- `horizon_ingest_`: performance measurements and stateful aspects of horizon's internal ingestion sub system
- `horizon_http_`: statistics and measurements of horizon's HTTP API service, all aspects of request/response load and timings. 
- `horizon_history_`: statistics on horizon ingested historical ledgers 
- `horizon_db_` 
- `process_`: generic host machine compute measurements
```

Author: oh, I had double line breaks in there on each group, but I think that was lost doing `npm run format:mdx`, thanks for reformat suggestion.

> 

Reviewer: Also, we need to add description about `horizon_db_`.

Author: oops, thanks for catching that, added that - https://github.com/stellar/stellar-docs/pull/198/commits/9fcae17acf12f491eff3a98a90700f5e74278d79

And for each key, there will be possibility of 0 or more labels, the serialized output(exposition) format follows this template:

```
<metrics_key_name>{label_1="value",label_2="value",,,} <value>
```

> Rather than listing all individual metrics keys in docs, as they change often, the recommendation is to perform an HTTP GET against the horizon metrics endpoint,`localhost:<ADMIN_PORT>\metrics`, using any http client(browser, curl, wget, etc) and the response will have the metrics keys and additional meta information on each metric key for description and type(counter, guage, histogram, summary), as an example for one key, `horizon_http_requests_duration_seconds`:

Reviewer: ```suggestion
Rather than listing all individual metrics keys in docs, as they change often, the recommendation is to perform an HTTP GET against the horizon metrics endpoint,`localhost:<ADMIN_PORT>/metrics`, using any http client(browser, curl, wget, etc) and the response will have the metrics keys and additional meta information on each metric key for description and type(counter, guage, histogram, summary), as an example for one key, `horizon_http_requests_duration_seconds`:
```

Author: there's no difference in the suggestion?

## Metrics

Metrics are emitted from Horizon over HTTP in [the de facto text based exposition format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format). The Metrics are published on the _private_ `/metrics` path of Horizon Admin port, which is an optional service to be started and will be bound by the horizon process onto the host machine loopback network(localhost or 127.0.0.1). To enable the Admin port, add environment configuration parameter `ADMIN_PORT=XXXXX`, the metrics endpoint will be reachable on the host machine as `localhost:<ADMIN_PORT>/metrics`. You can verify this by pointing any browser that can reach this address, it will print out all metrics keys.

### Exporting

> Once the admin port is enabled, the Horizon metrics endpoint can be 'scraped' by external monitoring infrastructure. Since the metrics output is encoded to [standard text based format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format) it will be compatibile for usage with many types of monitoring infrastructure that interoperate with same standard format.

Reviewer 2: ```suggestion
Once the admin port is enabled, the Horizon metrics endpoint can be 'scraped' by external monitoring infrastructure. Since the metrics output is encoded to [standard text based format](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format) it will be compatible for usage with many types of monitoring infrastructure that interoperate with the same standard format.
```

In the case of Horizon, metrics are published on the Admin http port which is bound to the host machine's loop back network interface(127.0.0.1), so external monitoring systems or services cannot reach the port directly. To expose the metrics securely, we recommend following the exporter pattern, which is common metrics scraping strategy.

One real world example for exporting from a bare metal horizon installation (horizon has been installed directly onto an operating system), use [FluentBit and the Prometheus Exporter](https://docs.fluentbit.io/manual/pipeline/outputs/prometheus-exporter) on the same host machine as Horizon is running. FluentBit will perform a simple port forwarding pipeline on the host machine. Configure the input to be Horizon's `localhost:<ADMIN_PORT>/metrics` and the output to be the `host` and `port` representing the target network interface and port on the host machine, and then configure your monitoring infrastructure to scrape that host address.

> In container orchestrated environents such as kubernetes, you can use the same exporter strategy. We assume you already have a metrics infrastructure deployment like Prometheus and Grafana setup on the cluster via the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator), and will just need to configure that infrastructure to scrape the horizon pod based on the ADMIN_PORT.

Reviewer 2: ```suggestion
In container orchestrated environments such as Kubernetes, you can use the same exporter strategy. We assume you already have a metrics infrastructure deployment like Prometheus and Grafana setup on the cluster via the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator), and will just need to configure that infrastructure to scrape the horizon pod based on the ADMIN_PORT.
```

### Data model

There are numerous application metrics keys emitted by horizon at runtime, encoded in four types of exposition formats: `counter`, `gauge`, `histogram`, `summary`. Each key is further qualified with labels for more granularity. To summarize, we can highlight the groupings of metrics keys by common function denoted in the prefix of their name:

> - `go_`: golang specfici runtime performance

Reviewer 2: ```suggestion
- `go_`: golang specfic runtime performance
```

- `horizon_txsub_`: attributes of horizon transaction submission sub system if enabled.

> - `horizon_stellar_core_`: runtime attributes of stellar network reported by the captive core or a remote stellar core which horizon is configured.

Author: good catch, removed.

- `horizon_order_book_`: runtime attributes of the in memory order book maintained by horizon of the current stellar network
- `horizon_log_`: counters of how many log messages printed at each severity level
- `horizon_ingest_`: performance measurements and stateful aspects of horizon's internal ingestion sub system
- `horizon_http_`: statistics and measurements of horizon's HTTP API service, all aspects of request/response load and timings.
- `horizon_history_`: statistics on horizon ingested historical ledgers
- `horizon_db_`: measurements on database performance, query times per endpoints, pooling stats

> - `process_`: generic host machine compute measurements

Reviewer 2: Should I be seeing metrics with "process" in the name?  I can't find it when I look at the metrics endpoint of my local Horizon.

Author: I think so, I saw the following in response:

```
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 5206.53
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 8192
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 49
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 1.196032e+08
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.69142892499e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.110106112e+09
# HELP process_virtual_memory_max_bytes Maximum amount of virtual memory available in bytes.
# TYPE process_virtual_memory_max_bytes gauge
process_virtual_memory_max_bytes -1
```

```
<metrics_key_name>{label_1="value",label_2="value",,,} <value>
```

Rather than listing all individual metrics keys in docs, as they change often, the recommendation is to perform an HTTP GET against the horizon metrics endpoint,`localhost:<ADMIN_PORT>/metrics`, using any http client(browser, curl, wget, etc) and the response will have the metrics keys and additional meta information on each metric key for description and type(counter, guage, histogram, summary), as an example for one key, `horizon_http_requests_duration_seconds`:

```
# HELP horizon_http_requests_duration_seconds HTTP requests durations, sliding window = 10m
# TYPE horizon_http_requests_duration_seconds summary
horizon_http_requests_duration_seconds{method="GET",route="/",status="200",streaming="false",quantile="0.5"} 0.000186958
horizon_http_requests_duration_seconds{method="GET",route="/",status="200",streaming="false",quantile="0.9"} 0.00043625
horizon_http_requests_duration_seconds{method="GET",route="/",status="200",streaming="false",quantile="0.99"} 0.000645
...

```

### Queries

> Build queries against the metrics data model to highlight performance of a given Horizon deployment. Refer to Stellar's [Grafana Horizon Dashboard](https://grafana.com/grafana/dashboards/13793-stellar-horizon/) for examples of metrics queries to derive application performance:

Reviewer 2: I can't seem to find some of [these](https://github.com/stellar/stellar-docs/pull/198/files#diff-eaaac1bf290b3eed0ad14d5e03bac252f1474e499b377ae807e87ecb612969f3R64-R67) on the Grafana dashboard. Are these queries provided as examples, or can I find the actual queries somewhere?

Author: What I'm aware of was dowloading the dashboard .json from https://grafana.com/grafana/dashboards/13793-stellar-horizon/, then can copy the metrics queries defined on panels with titles that should align fairly closes to the example queries listed here, if there is a gap where queries in there don't overlap enough with the example queries listed here, there's no other published grafana dashboards, so, would probably just remove from examples, which ones did you notice?

Reviewer 2: I couldn't find these
- Number of rate-limited requests.
- List of rate-limited IPs.
- Unique IPs.
- The most popular SDKs/apps sending requests to a given Horizon node.
- Average ingestion time of a transaction

- Average ingestion time of a ledger.
- Average ingestion time of a transaction.

> Choose the (revisions tab)[https://grafana.com/grafana/dashboards/13793-stellar-horizon/?tab=revisions], and download the dashboard source file to have access to the grafana dashboard source code and metrics queries that build each panel in dashboards.

Reviewer: ```suggestion
Choose the [revisions tab](https://grafana.com/grafana/dashboards/13793-stellar-horizon/?tab=revisions), and download the dashboard source file to have access to the grafana dashboard source code and metrics queries that build each panel in dashboards.
```

Author: ha, nice catch on typo.

### Alerts

> Once queries are developed on a Grafana dashboard, it enables convenient follow-on step to add [alert rules](https://grafana.com/docs/grafana/latest/alerting/alerting-rules/create-grafana-managed-rule/) based on specific queries to trigger notifications when thresholds are exceeded.

Reviewer: Can we add some more examples in the table?

Author: yes, added one more for a wide area related to http errors - https://github.com/stellar/stellar-docs/pull/198/commits/9fcae17acf12f491eff3a98a90700f5e74278d79

## Logs

Horizon will output logs to operating system's standard out. It will log on all aspects of runtime including http requests and ingestion. Typically, there are very few `warn` or `error` severity level messages emitted. The default severity level logged in Horizon is configured to `LOG_LEVEL=info`, this environment configuration parameter can be set to one of `trace, debug, info, warn, error`. The verbosity of log output is inverse of the severity level chosen. I.e. for most verbose logs use 'trace', for least verbose logs use 'error'.

For production deployments, we recommend using the default severity setting of `info` level and choose a log capture strategy depending on the deployment.

- Bare-metal deployment direct to operating system, redirect the standard out from Horizon process to a file on disk and apply a log rotation tool on the file such as [logrotate](https://man7.org/linux/man-pages/man8/logrotate.8.html) to manage disk space usage.
- Orchestrated deployment on kubernetes, use an EFK/ELK stack on the cluster and it can be configured to capture the standard out from Horizon pod.

## Runtime Profiling

> Horizon is written in golang, therefore it has been enabled to optionally emit the golang runtime diagnostics and profiling output (pprof)[https://go.dev/doc/diagnostics]. The pprof http endpoints are hosted on horizon's admin http port, it can be enabled by adding environment configuration parameter `ADMIN_PORT=XXXXX`, since the admin port binding is disabled by default.

Reviewer: ```suggestion
Horizon is written in golang, therefore it has been enabled to optionally emit the golang runtime diagnostics and profiling output [pprof](https://go.dev/doc/diagnostics). The pprof http endpoints are hosted on horizon's admin http port, it can be enabled by adding environment configuration parameter `ADMIN_PORT=XXXXX`, since the admin port binding is disabled by default.
```

For production deployments, we recommend using the default severity setting of `info` level and choose a log capture strategy depending on the deployment.

- Bare-metal deployment direct to operating system, redirect the standard out from Horizon process to a file on disk and apply a log rotation tool on the file such as [logrotate](https://man7.org/linux/man-pages/man8/logrotate.8.html) to manage disk space usage.
- Orchestrated deployment on kubernetes, use an EFK/ELK stack on the cluster and it can be configured to capture the standard out from Horizon pod.

## Runtime Profiling

Horizon is written in golang, therefore it has been enabled to optionally emit the golang runtime diagnostics and profiling output [pprof](https://go.dev/doc/diagnostics). The pprof http endpoints are hosted on horizon's admin http port, it can be enabled by adding environment configuration parameter `ADMIN_PORT=XXXXX`, since the admin port binding is disabled by default.

Two of the standard predefined profiles are published:

`localhost:<ADMIN_PORT>/debug/pprof/heap` - heap profiling

`localhost:<ADMIN_PORT>/debug/pprof/profile` - cpu profiling

> Use go's pprof command line tool to access the published endpoints and visualize the profiled dianostic data that is emitted. A brief example usage of the pprof tool from command line to get started, using `web` to display a graphical representation of current heap allocations:

Reviewer 2: ```suggestion
Use go's pprof command line tool to access the published endpoints and visualize the profiled diagnostic data that is emitted. A brief example usage of the pprof tool from command line to get started, using `web` to display a graphical representation of current heap allocations:
```

```

> $ go tool pprof http://localhost:6060/debug/pprof/heap

Reviewer 2: nice!
